### PR TITLE
feat: implement raven watcher for single object, add proper properties

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: Raven.NET.Core/Extensions/RavenSubjectExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        /// <summary>
+        /// Create string based on object reflection to store current value in memory cache
+        /// </summary>
+        /// <param name="obj"></param>
+        /// <param name="propName"></param>
+        /// <returns></returns>
+        public static string CreateCacheValue(this object obj, string propName = null)
+        {
+            var sb = new StringBuilder();
+            if (obj.GetType().IsValueType || obj is string)
```
Comment (pswierad-src) [inline Raven.NET.Core/Extensions/RavenSubjectExtensions.cs:17]: I'm more in favor of having if statements with brackets. Makes code a bit easier to read

File: Raven.NET.Core/Observers/Interfaces/IRavenWatcher.cs [hunk 2]
```diff
@@ -1,7 +1,14 @@
+using System;
+using Raven.NET.Core.Configuration;
+using Raven.NET.Core.Subjects;
+
 namespace Raven.NET.Core.Observers.Interfaces
 {
     public interface IRavenWatcher : IRaven
     {
-        
+        public IRavenWatcher Create(string name, Func<RavenSubject, bool> callback, Action<RavenSettings> options = null);
```
Comment (pswierad-src) [inline Raven.NET.Core/Observers/Interfaces/IRavenWatcher.cs:19]: Some documentation or inherit doc could be useful here. Just for reference

File: Raven.NET.Core/Observers/RavenTypeWatcher.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+using Raven.NET.Core.Observers.Interfaces;
+using Raven.NET.Core.Subjects;
+
+namespace Raven.NET.Core.Observers
+{
+    public class RavenTypeWatcher : IRavenTypeWatcher
+    {
+        void IRaven.Update(RavenSubject subject)
```
Comment (pswierad-src) [inline Raven.NET.Core/Observers/RavenTypeWatcher.cs:11]: Explicit modifier would make code easier to read

File: Raven.NET.Core/Extensions/RavenSubjectExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                sb.AppendFormat("{0}_{1}|", propName, obj);
+            else
+                foreach (var prop in obj.GetType().GetProperties())
+                {
+                    if(prop.Name == "Observers") continue;
+                    if (typeof(IEnumerable<object>).IsAssignableFrom(prop.PropertyType))
+                    {
+                        var get = prop.GetGetMethod();
+                        if (!get.IsStatic && get.GetParameters().Length == 0)
```
Comment (pswierad-src) [inline Raven.NET.Core/Extensions/RavenSubjectExtensions.cs]: If's could be inverted to reduce code nesting

File: Raven.NET.Core/Observers/RavenWatcher.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            
+            if(raven._ravenSettings.AutoDestroy)
+                TryDestroy(raven, name);
+        }
+
+        private void TryDestroy(RavenWatcher ravenWatcher, string ravenName)
+        {
+            if (!ravenWatcher.watchedSubjects.Any())
+            {
+                RavenCache.RavenWatcherCache.Remove(ravenName, out var _);
```
Comment (pswierad-src) [inline Raven.NET.Core/Observers/RavenWatcher.cs]: "var" is not needed when discarding out parameter

**Review by pswierad-src — CHANGES_REQUESTED** [review]
Some minor changes could be applied. Also some more cosmetic fixes could make code easier to read. Let me know what you think

File: Raven.NET.Core/Extensions/RavenSubjectExtensions.cs (hunk 1 above)
Comment (wisedev-code) [reply Raven.NET.Core/Extensions/RavenSubjectExtensions.cs:17]: I guess for me inline is habbit :D haha

File: Raven.NET.Core/Observers/Interfaces/IRavenWatcher.cs (hunk 2 above)
Comment (wisedev-code) [reply Raven.NET.Core/Observers/Interfaces/IRavenWatcher.cs:19]: good point

File: Raven.NET.Core/Observers/RavenWatcher.cs (hunk 5 above)
Comment (wisedev-code) [reply Raven.NET.Core/Observers/RavenWatcher.cs]: nice, didnt know about that

File: Raven.NET.Core/Observers/RavenTypeWatcher.cs (hunk 3 above)
Comment (wisedev-code) [reply Raven.NET.Core/Observers/RavenTypeWatcher.cs:11]: You can not set explicit modifer becouse this is not direct implementation of interface

Comment (pswierad-src) [reply Raven.NET.Core/Observers/RavenTypeWatcher.cs:11]: In that case it's fine

**Review by pswierad-src — APPROVED** [review]

**Pull request merged**